8632.6 combine ledger fix tests add documentation and tidy up (#8717)
msupply-foundation/open-msupply: server/service/src/ledger_fix/README.md

# Ledger Fix

Due to migration from existing mSupply sites and omSupply bug we've had ledger discrepancies, which are:
* Where total movements of stock (invoice_lines) does not add up to current stock balance (stock_line)

> * Where as any point in history, sum of invoice_lines as negative.

Reviewer: ```suggestion
* Where at any point in history, sum of invoice_lines as negative.
```

### Strategy to deal with ledger discrepancies

#### Prevent

Make it harder to implement things that will cause ledger discrepancy.

* Emphasise tests
* Service/repo structure with one point of entry for operations that affect stock

#### Fix and Notify

We understand that some ledger problems will sneak through (even with our best effort at prevention), to deal with affect of ledger discrepancies on users we will

* Implement a strategy of timely awareness of ledger discrepancies
* Strategy to fix ledger discrepancies in as much of a safe manner as possible

Logic in this section of our codebase is related to fix and notify.

### Ledger driver

There is a scheduler that would attempt to run, in the background

* Once a day
* After initialisation

> * After upgrade to this version

Reviewer: Just make it clearer when this was introduced in the far future

```suggestion
* After upgrading from a version < v2.9.2 to a version >= v2.9.2
```

This scheduler will try to find ledger discrepancies and then fix them. Scheduler last execution period is stored in key value store. Detailed logging for each stock line ledger fix would be added to normal file/console log and system log (to be synced to omSupply central), notifications can be made to support and dev team via system log (dashboard/notify) 

### Find Ledger Discrepancies

This operations should be quick and efficient, using partition sql introduced in stock_line/item ledger view, for each stock_line:

* Find any negative running balances 
* Find any cases where final running balances does not add up to total quantity and total quantity is not available quantity + any stock that is reserved but not yet applied to total

### Fix Ledger discrepancies

> Each stock line will go through a series of fixes based on know 'states' of ledger. After each fix a ledger discrepancy check is performed, if ledger is not yet fixed the next fix is attempted.

Reviewer: ```suggestion
Each stock line will go through a series of fixes based on known invalid of ledger states. After each fix a ledger discrepancy check is performed, if ledger is not yet fixed the next fix is attempted.

If a ledger cannot be corrected by the implemented automatic fixes, then our team should address them with after automated notification of their occurance.
```
